chore(husky): align pre-commit Python checks with CI

adityai0 opened the pull request:
**What kind of change does this PR introduce?**

This PR improves developer tooling by aligning local pre-commit checks with CI Python checks to fail fast before pushing.


**Issue Number:**

Fixes #5891 

**Snapshots/Videos:**

**N/A**

**If relevant, did you update the documentation?**

Yes
`docs/docs/docs/getting-started/installation.md `

**Summary**

This PR aligns local Husky pre-commit checks with the Python checks already enforced in CI to ensure contributors catch issues before pushing code.

Previously, missing Python dependencies or script failures were only detected in GitHub Actions, leading to avoidable CI failures and slower feedback. This change introduces local validation for the same Python scripts and documents the required setup so contributors can fail fast and iterate locally.

**Does this PR introduce a breaking change?**

No

## Checklist

### CodeRabbit AI Review
- [x] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [x] I have implemented or provided justification for each non-critical suggestion
- [x] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [x] I have written tests for all new changes/features
- [x] I have verified that test coverage meets or exceeds 95%
- [x] I have run the test suite locally and all tests pass


**Other information**

By running the same Python checks locally via Husky, this change reduces unnecessary GitHub Actions runs caused by easily catchable issues. This helps save CI minutes, avoids wasted compute resources, and shortens the feedback loop for contributors.

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Improvements**
  * Pre-commit checks now run via the project virtual environment and focus on staged files for consistent validation.
  * Exclusion handling can load exclusions from a file and normalizes paths for more accurate matching.

* **Chores**
  * Added a helper script to locate/prepare the project virtual environment.
  * Expanded the exclusions list used by line-counting workflows.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

adityai0 commented on .github/workflows/scripts/check_python_deps.py:
```diff
@@ -0,0 +1,68 @@
+
+    Returns:
+        None
+    """
```
@coderabbitai it fails locally and specifically mentions to add it

adityai0 commented:
@coderabbitai
full review using this prompt

https://github.com/PalisadoesFoundation/talawa-admin/pull/5416#issuecomment-3685222809

adityai0 commented on .github/workflows/scripts/test/test_check_python_deps.py:
```diff
@@ -0,0 +1,97 @@
+)
+
+if SPEC is None or SPEC.loader is None:
+    raise ImportError(f"Could not load check_python_deps.py from {SCRIPT_DIR}")
```
done

adityai0 commented:
@coderabbitai

full review using this prompt

https://github.com/PalisadoesFoundation/talawa-admin/pull/5416#issuecomment-3685222809

i have addressed all the issued and requested changes

adityai0 commented on .github/workflows/scripts/test/test_check_python_deps.py:
```diff
@@ -0,0 +1,131 @@
+            mock_pkg.metadata = {"Name": "existing-package"}
+            mock_dist.return_value = [mock_pkg]
+            result = check_python_deps.get_missing_dependencies()
+            self.assertIsInstance(result, list)
```
done

adityai0 commented:
I ran into an issue while merging the latest upstream changes, which caused some unintended file changes. The branch has now been reset and rebased cleanly on develop, keeping only the relevant Husky and CI parity updates. Apologies for the earlier noise.

palisadoes reviewed (changes requested):
@adityai0 
 
1. The issue is addressing the fact that .husky/pre-commit script does not run any of the CI/CD python checks.
4. This means that many times the PRs fail. For example the lack of translations or the presence of banned strings do not get flagged in the pre-commit stage only to fail in the PR.
5. How does this fix the issue?

palisadoes reviewed (changes requested):
1. How are the python packages installed? Some of the scripts have non-standard ones. You may need to setup a venv environment to make this work as part of the pre-commit.
2. Please update countline.py to take an argument of a text file with the files to exclude. That will make it easier as we can put the file in the `.github/workflows/config/` directory

adityai0 commented:
1. Currently, the Python packages must be installed manually by contributors. If any required package is missing, the Husky pre-commit hook fails and logs the missing dependencies using the `check_python_deps.py` script.
2. We can improve this by using a  `venv` to run all CI/CD Python checks. This would allow us to install and manage the required Python packages inside the `venv` instead of relying on global installations.
3. If this approach is fine, I will proceed with updating `check_python_deps.py` to first check whether a `venv` exists. If missing, create one (cross-platform) and install all required dependencies inside it.
4. I was also thinking about an additional improvement that we could create a single Python script that accepts `STAGED_SRC` and `CSS_STAGED` as arguments and executes all the Python CI checks. This would keep the Husky pre-commit hook minimal and readable. Additionally, if we add more CI checks in the future, they can be added into this script with least modification. Please let me know if this approach is acceptable.
5. For `countline.py`, I am working on updating the script to accept a text file as an argument.

palisadoes commented:
> 1. Currently, the Python packages must be installed manually by contributors. If any required package is missing, the Husky pre-commit hook fails and logs the missing dependencies using the `check_python_deps.py` script.
> 
>     2. We can improve this by using a  `venv` to run all CI/CD Python checks. This would allow us to install and manage the required Python packages inside the `venv` instead of relying on global installations.
> 
>     3. If this approach is fine, I will proceed with updating `check_python_deps.py` to first check whether a `venv` exists. If missing, create one (cross-platform) and install all required dependencies inside it.
> 
>     4. I was also thinking about an additional improvement that we could create a single Python script that accepts `STAGED_SRC` and `CSS_STAGED` as arguments and executes all the Python CI checks. This would keep the Husky pre-commit hook minimal and readable. Additionally, if we add more CI checks in the future, they can be added into this script with least modification. Please let me know if this approach is acceptable.
> 
>     5. For `countline.py`, I am working on updating the script to accept a text file as an argument.

1. You can just run the pip install with each invocation of the pre-commit and exit if venv isn't installed. Creating a python script that may require packages to be preinstalled to check for other packages doesn't seem a suitable approach.
2. Some of the centralized scripts may need to be updated to include a --files flag to make things easier.
3. Coming to think of it the centralization may not work if the scripts are located on the web. Can you think of other approaches to reduce duplication?

adityai0 commented:
Thanks for the insights, this helps clarify the direction.

1. I have reviewed the referenced PR from `talawa-api` to understand the approach being used there.
2. I’m thinking about a solution that aligns better with the issue and avoids unnecessary duplication.

I will take some time to refine this and come back tomorrow with a better solution. Thanks again for the guidance.

adityai0 commented on .github/workflows/scripts/venv.sh:
```diff
@@ -0,0 +1,36 @@
+else
+    echo "Error: Python is not installed."
+    exit 1
+fi
```
got it!

adityai0 commented on .github/workflows/scripts/venv.sh:
```diff
@@ -0,0 +1,26 @@
+# Install deps
+"$VENV_PY" -m pip install -q -r "$REPO_ROOT/.github/workflows/requirements.txt"
+
+echo "$VENV_PY"
```
done

adityai0 commented on .github/workflows/scripts/venv.sh:
```diff
@@ -0,0 +1,26 @@
+fi
+
+# Install deps
+"$VENV_PY" -m pip install -q --disable-pip-version-check -r "$REPO_ROOT/.github/workflows/requirements.txt" 2>/dev/null || true
```
working on it

adityai0 commented:
@coderabbitai
full review using this prompt #5416 (comment) 
i have addressed all the issues and requested changes.

adityai0 commented:
@coderabbitai
Do we need to add tests for this script, given that its behavior depends on the contributor’s local environment (Python/venv availability) and is primarily a thin shell wrapper rather than application logic?

adityai0 commented on .github/workflows/scripts/venv.sh:
```diff
@@ -0,0 +1,26 @@
+
+set -e
+
+REPO_ROOT=$(git rev-parse --show-toplevel)
```
done

adityai0 commented on .github/workflows/scripts/venv.sh:
```diff
@@ -0,0 +1,26 @@
+# Install deps
+"$VENV_PY" -m pip install -q --disable-pip-version-check -r "$REPO_ROOT/.github/workflows/requirements.txt"
+
+echo "$VENV_PY"
```
done

adityai0 commented on .github/workflows/scripts/venv.sh:
```diff
@@ -0,0 +1,26 @@
+else
+  echo "Error: Python executable not found in venv."
+  exit 1
+fi
```
done

adityai0 commented:
@coderabbitai
full review using this prompt https://github.com/PalisadoesFoundation/talawa-admin/pull/5416 (comment)
i have addressed all the issues and requested changes.
